akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -103,33 +105,65 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
-
-        // TODO consume the HTTP2-Settings header
-        // TODO consume the request entity and make sure it is inserted as an HTTP2 request
-        // https://http2.github.io/http2-spec/#rfc.section.3.2
-
-        Future.successful(
-          HttpResponse(
-            StatusCodes.SwitchingProtocols,
-            immutable.Seq[HttpHeader](
-              ConnectionUpgradeHeader,
-              UpgradeHeader,
-              // TODO add attributes from the request
-              UpgradeToOtherProtocolHeader(serverLayer)
+
+        log.debug("Got h2c upgrade request from HTTP/1.1 to HTTP2")
+
+        // https://http2.github.io/http2-spec/#Http2SettingsHeader 3.2.1 HTTP2-Settings Header Field
+        val upgradeSettings = req.headers.collect {
+          case raw: RawHeader if raw.lowercaseName == Http2SettingsHeader.name ⇒
+            Http2SettingsHeader.parse(raw.value)
+        }
+
+        upgradeSettings match {
+          // Must be exactly one
+          case immutable.Seq(Success(settingsFromHeader)) ⇒
+            // TODO remove duplication?
+
+            val injectedRequest =
+              if (req.method != HttpMethods.OPTIONS)
+                // inject the actual upgrade request with a stream identifier of 1
+                // https://http2.github.io/http2-spec/#rfc.section.3.2
+                Source.single(req.addHeader(Http2StreamIdHeader(1)))
+              else {
+                // No 100% sure about this but from the spec:
+                // If concurrency of an initial request with subsequent requests is important, an OPTIONS request
+                // can be used to perform the upgrade to HTTP/2, at the cost of an additional round trip.
+                req.discardEntityBytes()
+                Source.empty[HttpRequest]
+              }
+
+            val serverLayer: Flow[ByteString, ByteString, Future[Done]] = Flow.fromGraph(
+              Flow[HttpRequest]
+                .watchTermination()(Keep.right)
+                .merge(injectedRequest)
+                .via(Http2Blueprint.handleWithStreamIdHeader(parallelism)(handler)(system.dispatcher))
+                // the settings from the header are injected into the blueprint as initial demuxer settings
+                //
+                .joinMat(Http2Blueprint.serverStack(settings, log, settingsFromHeader))(Keep.left))
+
+            Future.successful(
+              HttpResponse(
+                StatusCodes.SwitchingProtocols,
+                immutable.Seq[HttpHeader](
+                  ConnectionUpgradeHeader,
+                  UpgradeHeader,
+                  UpgradeToOtherProtocolHeader(serverLayer)
+                )
+              )
             )
-          )
-        )
+
+          case _ ⇒
+
+            // A server MUST NOT upgrade the connection to HTTP/2 if this header field is not present or if more than one is present
+            log.debug("Invalid upgrade request (http2-settings header missing or repeated)")
```

Comment:
Perhaps it should still return `handler(req)` though?

Reply:
I think it'd make sense to log something (debug seems fine, perhaps info), but continue to handle the request as a normal non-http2 request.